### PR TITLE
tls: set failure reason in syscall error cases

### PR DESCRIPTION
**Pull request by mattklein123**
We lost most accounting of failure reasons when we switched to the
custom BIO.

Also, set failure reasons in some raw connection error cases while
here.

Risk Level: Low
Testing: New UTs
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: source/common/network/connection_impl.cc [hunk 1]
```diff
@@ -761,6 +762,11 @@ bool ConnectionImpl::bothSidesHalfClosed() {
 }
 
 absl::string_view ConnectionImpl::transportFailureReason() const {
+  // TODO(mattklein123): Is it possible for there to be a local failure reason and a transport
```
Comment (ggreenway) [inline source/common/network/connection_impl.cc:765]: Maybe make it an ENVOY_BUG if both are set, so we at least know that this is happening.

File: test/extensions/transport_sockets/tls/io_handle_bio_test.cc [hunk 2]
```diff
@@ -24,6 +25,15 @@ class IoHandleBioTest : public testing::Test {
   NiceMock<Network::MockIoHandle> io_handle_;
 };
 
+TEST_F(IoHandleBioTest, WriteError) {
+  EXPECT_CALL(io_handle_, writev(_, 1))
+      .WillOnce(Return(testing::ByMove(
+          Api::IoCallUint64Result(0, Api::IoErrorPtr(new Network::IoSocketError(100),
+                                                     Network::IoSocketError::deleteIoError)))));
+  EXPECT_EQ(-1, bio_->method->bwrite(bio_, nullptr, 10));
+  EXPECT_EQ(ERR_GET_LIB(ERR_get_error()), ERR_LIB_SYS);
```
Comment (ggreenway) [inline test/extensions/transport_sockets/tls/io_handle_bio_test.cc:34]: Also validate the value of `getSystemErrorCode()`

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall. 

/wait

File: test/extensions/transport_sockets/tls/io_handle_bio_test.cc (hunk 2 above)
Comment (mattklein123) [reply test/extensions/transport_sockets/tls/io_handle_bio_test.cc:34]: AFAICT there is no way to get the integer reason back out. I'm probably missing it? I will keep looking.

Comment (mattklein123) [reply test/extensions/transport_sockets/tls/io_handle_bio_test.cc:34]: Oh nevermind sorry I figured it out it's part of the packed error.

**Review by ggreenway — APPROVED** [review]

**Comment by rgs1** [comment]
@mattklein123 we pushed this out this morning and although we did not see a drop in 2xxs, we did see the connection_error SSL counter go up. Is it possible that these new `ERR_put_error()` calls:

```
    } else {
      ERR_put_error(ERR_LIB_SYS, 0, result.err_->getSystemErrorCode(), __FILE__, __LINE__);
```

are causing a false positive increment on:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/tls/ssl_socket.cc#L215

when they shouldn't?

**Comment by ggreenway** [comment]
@rgs1 can you tell what the error code was?

**Comment by rgs1** [comment]
@ggreenway I think:

```
[external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:212] [C2240416] TLS error: 33554536:system library:OPENSSL_internal:Connection reset by peer
```

**Comment by rgs1** [comment]
Does it sound like something that should be treated as `saw_counted_error` as well in:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/tls/ssl_socket.cc#L198

?

I also checked our active/destroy downstream connections and everything was healthy and following a similar pattern as before the deploy.

**Comment by rgs1** [comment]
FWIW, we are reverting internally for now since we are pretty confident that the increments in `ssl.connection_error` are not real.

**Comment by mattklein123** [comment]
OK thanks for the report I will look into this later this week. If we need to revert this change go ahead.

**Pull request merged**